File: health_prompt/HealthDashboard/Views/HDMoodTrendView.m
//
//  HDMoodTrendView.m
//  HealthDashboard - 心情趋势折线图
//
//  Created by zhang, haizi on 2026/3/19.
//

#import "HDMoodTrendView.h"

@implementation HDMoodTrendView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) { self.backgroundColor = [UIColor clearColor]; }
    return self;
}

- (void)reloadData { [self setNeedsDisplay]; }

- (void)drawRect:(CGRect)rect {
    if (!_records.count) return;
    NSInteger count = MIN(7, (NSInteger)_records.count);
    NSInteger start = _records.count - count;
    CGFloat w = rect.size.width;
    CGFloat h = rect.size.height;
    CGFloat stepX = w / (count - 1 > 0 ? count - 1 : 1);

    // 绘制折线
    UIBezierPath *linePath = [UIBezierPath bezierPath];
    linePath.lineWidth = 2;
    [[UIColor colorWithRed:0.38 green:0.75 blue:0.95 alpha:1.0] setStroke];

    NSMutableArray *points = [NSMutableArray array];
    for (NSInteger i = 0; i < count; i++) {
        HDMoodRecord *r = _records[start + i];
        CGFloat x = i * stepX;
        CGFloat y = h - (r.moodLevel / 5.0) * (h - 16) - 8;
        CGPoint pt = CGPointMake(x, y);
        [points addObject:[NSValue valueWithCGPoint:pt]];
        if (i == 0) [linePath moveToPoint:pt];
        else        [linePath addLineToPoint:pt];
    }
    [linePath stroke];

    // 绘制节点和表情
    for (NSInteger i = 0; i < count; i++) {
        HDMoodRecord *r = _records[start + i];
        CGPoint pt = [points[i] CGPointValue];

        // 节点圆
        UIBezierPath *dot = [UIBezierPath bezierPathWithArcCenter:pt radius:4 startAngle:0 endAngle:2*M_PI clockwise:YES];
        [[UIColor colorWithRed:0.22 green:0.54 blue:0.95 alpha:1.0] setFill];
        [dot fill];

        // 表情文字（最后一个节点显示）
        if (i == count - 1) {
            NSString *emoji = r.emojiString;
            [emoji drawAtPoint:CGPointMake(pt.x - 8, pt.y - 22)
                withAttributes:@{NSFontAttributeName: [UIFont systemFontOfSize:16]}];
        }
    }
}

@end

File: health_advanced_swift_spec/HealthDashboard/Views/HDMoodTrendView.swift
//
//  HDMoodTrendView.swift
//  HealthDashboard - 心情趋势折线图
//
//  Created by zhang, haizi on 2026/3/22.
//

import UIKit

/// 心情趋势折线图
class HDMoodTrendView: UIView {

    // MARK: - Properties

    var records: [HDMoodRecord] = []

    // MARK: - Init

    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .clear
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }

    // MARK: - Public

    func reloadData() {
        setNeedsDisplay()
    }

    // MARK: - Drawing

    override func draw(_ rect: CGRect) {
        guard !records.isEmpty else { return }
        let count = min(7, records.count)
        let start = records.count - count
        let w = rect.width
        let h = rect.height
        let stepX = w / CGFloat(max(count - 1, 1))

        // 计算节点坐标
        var points: [CGPoint] = []
        for i in 0..<count {
            let r = records[start + i]
            let x = CGFloat(i) * stepX
            let y = h - (CGFloat(r.moodLevel) / 5.0) * (h - 16) - 8
            points.append(CGPoint(x: x, y: y))
        }

        // 绘制折线
        let linePath = UIBezierPath()
        linePath.lineWidth = 2
        UIColor(red: 0.38, green: 0.75, blue: 0.95, alpha: 1).setStroke()
        for (i, pt) in points.enumerated() {
            if i == 0 { linePath.move(to: pt) }
            else { linePath.addLine(to: pt) }
        }
        linePath.stroke()

        // 绘制节点
        UIColor(red: 0.22, green: 0.54, blue: 0.95, alpha: 1).setFill()
        for (i, pt) in points.enumerated() {
            let dot = UIBezierPath(arcCenter: pt, radius: 4, startAngle: 0, endAngle: 2 * .pi, clockwise: true)
            dot.fill()

            // 最后一个节点显示 emoji
            if i == count - 1 {
                let r = records[start + i]
                let emoji = r.emojiString as NSString
                emoji.draw(at: CGPoint(x: pt.x - 8, y: pt.y - 22),
                           withAttributes: [.font: UIFont.systemFont(ofSize: 16)])
            }
        }
    }
}
